openwillis/measures/audio/speech_preprocess.py

```diff
@@ -0,0 +1,145 @@
+
+    ------------------------------------------------------------------------------------------------------
+    """
+
+    num_channels = audio_signal.channels
+
+    if num_channels == 1:
+        offset = audio_signal.get_dc_offset(channel=1)
+        if offset == 0:
+            return audio_signal
+        audio_signal.remove_dc_offset(channel=1, offset=offset)
+    else:
+        offset_l = audio_signal.get_dc_offset(channel=1)
+        offset_r = audio_signal.get_dc_offset(channel=2)
+        offset = (offset_l + offset_r) / 2
+
+        if offset == 0:
+            return audio_signal
+
+        audio_signal.remove_dc_offset(channel=1, offset=offset_l)
+        audio_signal.remove_dc_offset(channel=2, offset=offset_r)
+
+    return audio_signal
+
+def volume_normalization(audio_signal, target_dBFS):
+    """
+    ------------------------------------------------------------------------------------------------------
+    
+    Normalizes the volume of the audio signal to the target dBFS.
+    
+    Parameters:
+    ...........
+    audio_signal : pydub.AudioSegment
+        input audio signal
+    target_dBFS : float
+        target dBFS
+        
+    Returns:
+    ...........
+    pydub.AudioSegment
+        normalized audio signal
+
+    ------------------------------------------------------------------------------------------------------
+    """
+
+    headroom = -audio_signal.max_dBFS
+    gain_adjustment = target_dBFS - audio_signal.dBFS
+
+    if gain_adjustment > headroom:
+        gain_adjustment = headroom
+
+    audio_signal = audio_signal.apply_gain(gain_adjustment)
+    return audio_signal
+
+def audio_preprocess(audio_in, audio_out):
+    """
+    ------------------------------------------------------------------------------------------------------
+
+    Preprocesses audio signal by resampling, removing DC offset, normalizing volume, and denoising.
+
+    Parameters:
+    ...........
+    audio_in : str
+        path to the input audio file
+    audio_out: str
+        path to the output audio file
+
+    Returns:
+    ...........
+    None
+
+    ------------------------------------------------------------------------------------------------------
+    """
+
+
+    try:
+        if not audio_in.endswith(".wav") and not audio_in.endswith(".mp3"):
+            logger.error(f'Error in audio preprocessing- file: {audio_in} & Error: File format not supported')
+            return
+        
+        if not os.path.exists(audio_in):
+            logger.error(f'Error in audio preprocessing- file: {audio_in} & Error: File not found')
+            return
+
+        audio_signal = AudioSegment.from_file(audio_in, format="wav" if audio_in.endswith(".wav") else "mp3")
+
+        audio_signal = resample(audio_signal, 44100)
+        audio_signal = dc_offset(audio_signal)
+        audio_signal = volume_normalization(audio_signal, -20)
+
+        audio_signal.export(audio_out, format="wav" if audio_out.endswith(".wav") else "mp3")
```

Comment:
Can we also consider the usage of WillisPipeline? Upon successful execution, this function exports the audio file to the specified output path. However, in the case of WillisPipeline, it exports the file inside the container memory, which is unnecessary and not required. Since this container is intended to directly upload the cleaned signals to S3. 

To address this, could we implement a behavior similar to speaker separation, where the function simply returns the signal, and we use the "to_audio" function to convert the signals into an audio recording if needed?

Please check with Anzar on this.

Reply:
If he agrees, we are always going to return audio signals so return that in the finally block